### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by chrissearle** [javaBin/ems-ios#149]
New feedback service coming - we will likely need to add feedback posting this year. More info to follow.

**Comment by ingarabr** [comment]
The API is documented in the readme file here: https://github.com/javaBin/devnull

**Comment by chrissearle** [comment]
Simple URL built from event and session ID. Range 0-5.

We'll need the following:

* GUI
* Update core data to allow us to store - otherwise we can't show the user what they posted (since it's anonymous - there's no fetch option). This means adding a new migration too - remember to test it :)

When should feedback be opened? Earlier versions we had "Not until the last 10 mins of a session" to avoid people voting in advance without listening.

**Comment by ronnienessa** [comment]
Would be nice to separate favorites/rating etc from the sessions data model. That way it would be easier to handle migrations/and we could simply wipe the sessions on update and/or in case of problems.

**Comment by chrissearle** [comment]
Keeping it as a separate table makes the migration far simpler - "add this new object" is a fairly safe thing.

There will be a relationship tho - we need to know what session we're rating and/or displaying rating for.

**Comment by ronnienessa** [comment]
I'm not sure we actually need core data relationships for this. 
 What about keeping it in separate core data database? We could use the uri of the session as foreign key.

**Comment by chrissearle** [comment]
That might work.

I'd keep it in the same moc - but a completely separate object.

**Comment by chrissearle** [comment]
db: same moc - new object - with url key (the constructed url for posting based on the session) and current rating (if not present rating is 0).

GUI - new row on the detail view table with 5 buttons per rating (overall, relevance, content, quality) -  star marked - selected yellow, normal grey - perhaps with own row class to make connecting and working out the rating happen in one place? Going to be a lot of stars.

Own view? Might be easiest - but need to push to it from somewhere - again - row on detail view that pushes to rating view?

Save to moc and post - do we care enough to make it atomic (as in only save if post is OK)?

Do ratings need to be locked after entry to avoid double (or more) posts to server?

**Comment by ronnienessa** [comment]
You´re right, same moc seems right for ratings (and maybe even relation). favorites would be nicer to have in separate moc. That way the sessions database would be a discardable cache...

**Comment by ronnienessa** [comment]
Have a look at the WWDC app. For some reason it is very similar to ours... Weird.  They have a separate view controller for ratings.

**Comment by ronnienessa** [comment]
As for the atomic stuff. Wouldn´t it be better to refetch ratings for session after post? I´ve never liked that keeping things in sync manually. The server knows the correct answer...

**Comment by chrissearle** [comment]
Check the API - on the end of https://github.com/javaBin/devnull

**Comment by ronnienessa** [comment]
I saw that. It does not say anything about what it returns after post though. I assume the ratings will be part of the sessions. If thats correct and assuming the POST itself does not return anything useful I think it would be better to: 
1. POST rating
2. If OK refetch session or session rating -> save to moc

**Comment by chrissearle** [comment]
POST is OK - but there is no fetch - this doesn't currently send anything to EMS - so if we don't store what we sent then we won't know what we sent. So we store in the db what was entered and the post fails. Do we tidy up, warn the user or just let it go?

**Comment by ronnienessa** [comment]
Ah ok. It´s off course user specific.. Then I´m back to separate MOC. Procedure should probably be: 

1. Save MOC with users rating. 
2. On a low priority background queue schedule an operation that POST the users rating to the server. This could for instance be scheduled as a side effect of the store change itself or batched whenever the device feels its ok to run some background network calls. If it fails, we could retry with some exponential backoff algorithm… If it fails more than n number of times just discard the operation and log to whatever tracking tool we are currently using.. MOC does not need to be updated. 

This should work nicely with the offline nature of the app.

**Comment by ingarabr** [comment]
We can add a get method on the same endpoint as the post in the feedback api if it's needed

**Comment by chrissearle** [comment]
Returning the personal submission (we talked about NSUUID identifierForVendor as ID) then I don't really see the need. Might be nice if we could grab the current aggregate/average rating for a session tho - could have a 

Rating: <current average rating>  (>)

Where (>) is the apple disclosure indicator so the user knows that clicking allows them to enter their own rating.

This aggregate could be added to EMS or fetched from devnull (this changes when things happen - one on refresh of sessions - one on display of a session).

**Comment by chrissearle** [comment]
Mostly done. Need to only open when 5 mins of session left.
